src/core/processing/processingalgorithm.cpp

```diff
@@ -0,0 +1,271 @@
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+
+#include "processingalgorithm.h"
+
+#include <qgsapplication.h>
+#include <qgsprocessingalgorithm.h>
+#include <qgsprocessingprovider.h>
+#include <qgsprocessingregistry.h>
+#include <qgsvectorlayer.h>
+#include <qgsvectorlayerutils.h>
+
+
+ProcessingAlgorithm::ProcessingAlgorithm( QObject *parent )
+  : QObject( parent )
+{
+}
+
+void ProcessingAlgorithm::setId( const QString &id )
+{
+  if ( mAlgorithmId == id )
+  {
+    return;
+  }
+
+  mAlgorithmId = id;
+  mAlgorithm = !mAlgorithmId.isEmpty() ? QgsApplication::instance()->processingRegistry()->algorithmById( mAlgorithmId ) : nullptr;
+
+  emit idChanged( mAlgorithmId );
+
+  if ( mPreview && mAlgorithm )
+  {
+    run( true );
+  }
+}
+
+QString ProcessingAlgorithm::displayName() const
+{
+  return mAlgorithm ? mAlgorithm->displayName() : QString();
+}
+
+QString ProcessingAlgorithm::shortHelp() const
+{
+  return mAlgorithm ? mAlgorithm->shortHelpString() : QString();
+}
+
+void ProcessingAlgorithm::setInPlaceLayer( QgsVectorLayer *layer )
+{
+  if ( mInPlaceLayer.data() == layer )
+  {
+    return;
+  }
+
+  mInPlaceLayer = layer;
+  emit inPlaceLayerChanged();
+
+  if ( mPreview && mAlgorithm )
+  {
+    run( true );
+  }
+}
+
+void ProcessingAlgorithm::setInPlaceFeatures( const QList<QgsFeature> &features )
+{
+  if ( mInPlaceFeatures == features )
+  {
+    return;
+  }
+
+  mInPlaceFeatures = features;
+
+  emit inPlaceFeaturesChanged();
+
+  if ( mPreview && mAlgorithm )
+  {
+    run( true );
+  }
+}
+
+void ProcessingAlgorithm::setParameters( const QVariantMap &parameters )
+{
+  if ( mAlgorithmParameters == parameters )
+  {
+    return;
+  }
+
+  mAlgorithmParameters = parameters;
+
+  emit parametersChanged();
+
+  if ( mPreview && mAlgorithm )
+  {
+    run( true );
+  }
+}
+
+void ProcessingAlgorithm::setPreview( bool preview )
+{
+  if ( mPreview == preview )
+  {
+    return;
+  }
+
+  mPreview = preview;
+
+  emit previewChanged();
+
+  if ( mPreview && mAlgorithm )
+  {
+    run( true );
+  }
+  else
+  {
+    if ( !mPreviewGeometries.isEmpty() )
+    {
+      mPreviewGeometries.clear();
+      emit previewGeometriesChanged();
+    }
+  }
+}
+
+bool ProcessingAlgorithm::run( bool previewMode )
+{
+  if ( !mAlgorithm )
+  {
+    return false;
+  }
+
+  if ( previewMode )
+  {
+    mPreviewGeometries.clear();
+    emit previewGeometriesChanged();
+  }
+
+  QgsProcessingContext context;
+  context.setFlags( QgsProcessingContext::Flags() );
+  context.setProject( QgsProject::instance() );
+
```

Comment:
You might consider setting the various invalid geometry settings / callbacks for context too. These could be important for QField use, eg if someone has digitized an invalid polygon.